src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ExecuteActionTests.cs

```diff
@@ -0,0 +1,132 @@
+{
+    public sealed class ExecuteActionTests
+    {
+        private const int TokenTimeoutMs = 2000; // Must be less than DelayMs in the ExecuteApp
+
+        private const string TaskCanceledMessage = "A task was canceled";
+
+        [Fact]
+        public async Task ExecuteAction_ZeroExitCode()
+        {
+            ExecuteAction action = new();
+
+            ExecuteOptions options = new();
+
+            options.Path = DotNetHost.HostExePath;
+            options.Arguments = GenerateArgumentsString(new string[] { "ZeroExitCode" });
+
+            CollectionRuleActionResult result = await action.ExecuteAsync(options, null, CreateCancellationToken());
+
+            Assert.Equal("0", result.OutputValues["ExitCode"]);
+        }
+
+        [Fact]
+        public async Task ExecuteAction_NonzeroExitCode()
+        {
+            ExecuteAction action = new();
+
+            ExecuteOptions options = new();
+
+            options.Path = DotNetHost.HostExePath;
+            options.Arguments = GenerateArgumentsString(new string[] { "NonzeroExitCode" });
+
+            InvalidOperationException invalidOperationException = await Assert.ThrowsAsync<InvalidOperationException>(
+                () => action.ExecuteAsync(options, null, CreateCancellationToken()));
+
+            Assert.Contains(string.Format(Strings.ErrorMessage_NonzeroExitCode, "-1"), invalidOperationException.Message);
+        }
+
+        [Fact]
+        public async Task ExecuteAction_TokenCancellation()
+        {
+            ExecuteAction action = new();
+
+            ExecuteOptions options = new();
+
+            options.Path = DotNetHost.HostExePath;
+            options.Arguments = GenerateArgumentsString(new string[] { "TokenCancellation" }); ;
+
+            TaskCanceledException taskCanceledException = await Assert.ThrowsAsync<TaskCanceledException>(
+                () => action.ExecuteAsync(options, null, CreateCancellationToken()));
+
+            Assert.Contains(TaskCanceledMessage, taskCanceledException.Message);
+        }
+
+        [Fact]
+        public async Task ExecuteAction_TextFileOutput()
+        {
+            ExecuteAction action = new();
+
+            ExecuteOptions options = new();
+
+            DirectoryInfo outputDirectory = Directory.CreateDirectory(Path.Combine(Path.GetTempPath(), "ExecuteAction", Guid.NewGuid().ToString()));
+            string textFileOutputPath = outputDirectory.FullName + "\\file.txt";
+
+            const string testMessage = "TestMessage";
+
+            options.Path = DotNetHost.HostExePath;
+            options.Arguments = GenerateArgumentsString(new string[] { "TextFileOutput", textFileOutputPath, testMessage });
+
+            CollectionRuleActionResult result = await action.ExecuteAsync(options, null, CreateCancellationToken());
+
+            Assert.Equal("0", result.OutputValues["ExitCode"]);
+            Assert.Equal(testMessage, File.ReadAllText(textFileOutputPath));
+
+            try
```

Comment:
Most of the test should be wrapped in a try...finally where in the finally block you delete the directory.